lib/elixir/lib/access.ex

```diff
@@ -715,6 +715,42 @@ defmodule Access do
     {nil, :lists.reverse(updates)}
   end
 
+  @doc ~S"""
+  Same as `at/1` except that it raises `Access.OutOfBoundsError`
+  if the given index is out of bounds.
+
+  ## Examples
+
+      iex> get_in([:a, :b, :c], [Access.at!(2)])
+      :c
+      iex> get_in([:a, :b, :c], [Access.at!(3)])
+      ** (Access.OutOfBoundsError) out of bounds error
+
+  """
+  @spec at!(integer) :: access_fun(data :: list, get_value :: term)
+  def at!(index) when is_integer(index) do
+    fn op, data, next -> at!(op, data, index, next) end
+  end
+
+  defp at!(:get, data, index, next) when is_list(data) do
+    case Enum.fetch(data, index) do
+      {:ok, value} -> next.(value)
+      :error -> raise Access.OutOfBoundsError
+    end
+  end
+
+  defp at!(:get_and_update, data, index, next) when is_list(data) do
+    if index >= length(data) do
```

Comment:
Unfortunately this will traverse the data twice and it may not work for negative indexes. My suggestion is to allow an anonymous function to be given to `get_and_update_at` which will be called whenever it returns `nil`. For the regular case, the anonymous function will be `fn -> nil end` but for those case we will raise. WDYT?

Reply:
Good catch, and makes sense 👍🏻 Thanks!